IDinsight/aaq-core
# [AAQ-458] Add langfuse

Reviewer: @amiraliemami 
Estimate: 1 hour

---

## Ticket

Fixes: https://idinsight.atlassian.net/browse/AAQ-458

## Description

### Goal

Add tracing so that we can track input outputs of LLM calls. (see discussion [here](https://idinsight.atlassian.net/wiki/spaces/PD/pages/2543845378/Embeddings+LLM+Observability)

This is the implementation of the suggestion in the design doc.

### Changes

1. Pass user_id and query_id metadata to LiteLLM calls
2. Add option to use Langfuse for tracing (env vars, template.env, docs, ..)

https://www.loom.com/share/e2952bd4b4964aee906f8ad13587b322

### Future Tasks (optional)

[ ] Figure out how to add feedback using [Langfuse scoring](https://python.reference.langfuse.com/langfuse/client#Langfuse.score)
[ ] Env vars are getting messy! Separate them per app, and use Pydantic settings.

## How has this been tested?

1. Get your own Langfuse keys
2. Set new env vars as per docs / template.env in your .env file
3. Docker compose up
4. Log-in as any user (*not tested with Google)
5. Make /embeddings-search or /llm-response requests and check Langfuse console for new traces

## To-do before merge (optional)

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [ ] I have added blogpost in Latest Updates
